File: clients/ruby/cac-client/cacclient.rb
require 'ffi'
require 'os'
require 'dotenv'

Dotenv.load

module MergeStrategy
    REPLACE = "REPLACE"
    MERGE = "MERGE"
end

module CACLib
    extend FFI::Library

    extension = "dylib"
    if OS.windows?
        extension = "dll"
    elsif OS.mac?
        extension = "dylib"
    else
        extension = "so"
    end
    libpath = File.expand_path(File.join(__dir__, "#{ENV["SUPERPOSITION_LIB_PATH"]}/libcac_client.#{extension}"))
    ffi_lib libpath

    class ArcClient < FFI::Struct
        layout :dummy, :int
    end

    # Attach C functions
    attach_function :cac_last_error_length, [], :int
    attach_function :cac_last_error_message, [], :string
    attach_function :cac_free_string, [:pointer], :void
    attach_function :cac_new_client, [:string, :ulong, :string], :int
    attach_function :cac_start_polling_update, [:string], :void
    attach_function :cac_free_client, [ArcClient.by_ref], :void
    attach_function :cac_get_client, [:string], ArcClient.by_ref
    attach_function :cac_get_last_modified, [ArcClient.by_ref], :string
    attach_function :cac_get_config, [ArcClient.by_ref, :string, :string], :string
    attach_function :cac_get_resolved_config, [ArcClient.by_ref, :string, :string, :string], :string
    attach_function :cac_get_default_config, [ArcClient.by_ref, :string], :string
end

class CACClient
  def initialize(tenant, update_frequency, hostname)
    result = CACLib.cac_new_client(tenant, update_frequency, hostname)
    raise "Failed to create client: #{last_error_message}" if result != 0
    @client = CACLib.cac_get_client(tenant)
    raise "Failed to get client: #{last_error_message}" if @client.null?
    @tenant = tenant
  end

  def start_polling_update
    CACLib.cac_start_polling_update(@tenant)
  end

  def get_last_modified
    result = CACLib.cac_get_last_modified(@client)
    raise "Failed to get last modified time: #{last_error_message}" if result.nil?
    result
  end

  def get_config(filter_query, filter_prefix)
    result = CACLib.cac_get_config(@client, filter_query, filter_prefix)
    puts "Filter query  : #{result}"
    raise "Failed to get config: #{last_error_message}" if result.nil?
    result
  ensure
    # CACLib.cac_free_string(result) if result // No need to free as its handled internally
  end

  def get_resolved_config(query, filter_keys, merge_strategy)
    result = CACLib.cac_get_resolved_config(@client, query, filter_keys, merge_strategy)
    raise "Failed to get resolved config: #{last_error_message}" if result.nil?
    result
  end

  def get_default_config(filter_keys)
    result = CACLib.cac_get_default_config(@client, filter_keys)
    raise "Failed to get default config: #{last_error_message}" if result.nil?
    result
  end

  def last_error_message
    CACLib.cac_last_error_message
  end

  def close
    CACLib.cac_free_client(@client) unless @client.null?
  end
end

File: clients/legacy/java/cac-client/src/main/java/cac_client/CacClient.java
package cac_client;

import java.io.IOException;

import jnr.ffi.LibraryLoader;
import jnr.ffi.Pointer;

public class CacClient {

    public interface RustLib {
        String cac_last_error_message();

        int cac_new_client(String tenant, long updateFrequency, String hostName);
        
        int cac_new_client_with_cache_properties(String tenant, long updateFrequency, String hostName, long cacheMaxCapacity, long cacheTTL, long cacheTTI);

        void cac_free_client(Pointer ptr);

        Pointer cac_get_client(String tenant);

        Pointer cac_get_config(Pointer clientPtr, String filterQuery, String filterPrefix);

        Pointer cac_get_default_config(Pointer clientPtr, String filterKeys);

        void cac_start_polling_update(String tenant);

        Pointer cac_get_last_modified(Pointer clientPtr);

        Pointer cac_get_resolved_config(Pointer clientPtr, String filterQuery, String filterPrefix, String merge_strategy);

        void cac_free_string(Pointer s);
    }

    public static RustLib rustLib;

    public CacClient() {
        String libraryName = "cac_client";
        String libraryPath = System.getenv("SUPERPOSITION_LIB_PATH");
        System.out.println("libraryPath" + libraryPath);
        System.setProperty("jnr.ffi.library.path", libraryPath);
        CacClient.rustLib = LibraryLoader.create(RustLib.class).load(libraryName);
    }

    public int cacNewClient(String tenant, long updateFrequency, String hostName) throws CACClientException {
        
        int result = rustLib.cac_new_client(tenant, updateFrequency, hostName);
        if (result > 0) {
            String errorMessage = rustLib.cac_last_error_message();
            throw new CACClientException("Failed to create new CAC client: " + errorMessage);
        }
        return result;
    }

    public int cacNewClientWithCacheProperties(String tenant, long updateFrequency, String hostName, Long cacheMaxCapacity, Long cacheTTL, Long cacheTTI) throws CACClientException {
       
        
        int result = rustLib.cac_new_client_with_cache_properties(tenant, updateFrequency, hostName, cacheMaxCapacity, cacheTTL, cacheTTI);
        if (result > 0) {
            String errorMessage = rustLib.cac_last_error_message();
            throw new CACClientException("Failed to create new CAC client: " + errorMessage);
        }
        return result;
    }

    public Pointer getCacClient(String tenant) throws IOException {
        Pointer clientPtr = rustLib.cac_get_client(tenant);
        return clientPtr;
    }

    public String getConfig(Pointer clientPtr, String filterQuery, String filterPrefix) throws IOException {
        Pointer result = rustLib.cac_get_config(clientPtr, filterQuery, filterPrefix);
        if (result == null) {
            String errorMessage = rustLib.cac_last_error_message();
            throw new IOException("Failed to get config: " + errorMessage);
        }
        String config = getStringAndFree(result);
        return config;
    }

    public String getDefaultConfig(Pointer clientPtr, String filterKeys) throws IOException {
        Pointer result = rustLib.cac_get_default_config(clientPtr, filterKeys);
        if (result == null) {
            String errorMessage = rustLib.cac_last_error_message();
            throw new IOException("Failed to get default config: " + errorMessage);
        }
        String config = getStringAndFree(result);
        return config;
    }

    public void startPollingUpdate(String tenant) throws IOException {
        rustLib.cac_start_polling_update(tenant);
    }

    public String getLastModified(Pointer clientPtr) throws IOException {
        Pointer result = rustLib.cac_get_last_modified(clientPtr);
        if (result == null) {
            String errorMessage = rustLib.cac_last_error_message();
            throw new IOException("Failed to get last modified: " + errorMessage);
        }
        String lastModified = getStringAndFree(result);
        return lastModified;
    }

    public String getResolvedConfig(Pointer clientPtr, String filterQuery, String filterPrefix, String mergeStrategy)
            throws IOException {
        Pointer result = rustLib.cac_get_resolved_config(clientPtr, filterQuery, filterPrefix, mergeStrategy);
        if (result == null) {
            String errorMessage = rustLib.cac_last_error_message();
            throw new IOException("Failed to get resolved config: " + errorMessage);
        }
        String resolvedConfig = getStringAndFree(result);
        return resolvedConfig;
    }

    public void cacFreeClient(Pointer clientPtr) throws IOException{
        rustLib.cac_free_client(clientPtr);
    }

    public String getLastError() throws IOException {
        String errorMessage = rustLib.cac_last_error_message();
        if (errorMessage != null) {
            return errorMessage;
        } else {
            return "No error";
        }
    }

    public void freeString(Pointer ptr) {
        rustLib.cac_free_string(ptr);
    }

    public String getStringAndFree(Pointer ptr) {
        if (ptr == null) {
            return null;
        }
        try {
            return ptr.getString(0);
        } finally {
            freeString(ptr);
        }
    }
}
